sw/host/ot_certs/src/asn1/codegen.rs

```diff
@@ -0,0 +1,515 @@
+
+    /// Push a byte array of fixed length into the ASN1 output. Note that this function does not add a tag to
+    /// the ASN1 output.
+    fn push_byte_array(&mut self, name_hint: Option<String>, val: &Value<Vec<u8>>) -> Result<()> {
+        match val {
+            Value::Literal(x) => {
+                let const_name = self.add_constant_byte_array(name_hint, x);
+                self.push_str_with_indent(&format!("RETURN_IF_ERROR(asn1_push_bytes(&state, {const_name}, sizeof({const_name})));\n"));
+                // There is not tag, we are just pushing the data itself.
+                self.max_out_size += x.len();
+            }
+            Value::Variable(Variable { name, convert }) => {
+                let VariableInfo {
+                    codegen,
+                    var_type: source_type,
+                } = (self.variable_info)(name)?;
+                match source_type {
+                    VariableType::ByteArray { size } => {
+                        ensure!(convert.is_none(), "using a byte-array variable for a byte-array field cannot specify a conversion");
+                        let VariableCodegenInfo::Pointer {
+                            ptr_expr,
+                            size_expr,
+                        } = codegen
+                        else {
+                            bail!("internal error: byte-array represented by a VariableCodegenInfo::Int32");
+                        };
+                        // There is not tag, we are just pushing the data itself.
+                        self.max_out_size += size;
+                        self.push_str_with_indent(&format!(
+                            "RETURN_IF_ERROR(asn1_push_bytes(&state, {ptr_expr}, {size_expr}));\n"
+                        ))
+                    }
+                    _ => bail!(
+                        "using a variable of type {:?} for a byte-array field is not supported",
+                        source_type
+                    ),
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Push an optionally tagged string into the ASN1 output.
+    fn push_string(
+        &mut self,
+        _name_hint: Option<String>,
+        str_type: &Tag,
+        val: &Value<String>,
+    ) -> Result<()> {
+        let str_type = str_type.codestring();
+        match val {
+            Value::Literal(x) => {
+                // NOTE: we are not using a constant for the size of the string because technically there is no way
+                // to know what the `char` type is and therefore what is the length of this string for the C compiler.
+                self.push_str_with_indent(&format!("RETURN_IF_ERROR(asn1_push_string(&state, {str_type}, \"{x}\", strlen(\"{x}\")));\n"));
+                // A tagged string needs a tag, up to 3 bytes of length and the string itself.
+                // Don't try to exactly compute how many bytes we need for the length.
+                self.max_out_size += 1 + 3 + x.len();
```

Comment:
Nit: if the `1` is for the null terminator, can we put it at the end to make that clear?